kv/etcd/options.go

```diff
@@ -0,0 +1,167 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"errors"
+	"time"
+
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/retry"
+)
+
+var (
+	defaultRequestTimeout         = 10 * time.Second
+	defaultDialTimeout            = 10 * time.Second
+	defaultWatchChanCheckInterval = 10 * time.Second
+	defaultWatchChanResetInterval = 10 * time.Second
+	defaultKeyFn                  = KeyFn(
+		func(key string) string {
+			return key
+		},
+	)
+)
+
+// KeyFn is a function that wraps a key
+type KeyFn func(key string) string
+
+// Options are options for the client of the kv store
+type Options interface {
+	RequestTimeout() time.Duration
+
+	SetRequestTimeout(t time.Duration) Options
+
+	KeyFn() KeyFn
+
+	SetKeyFn(f KeyFn) Options
+
+	InstrumentsOptions() instrument.Options
+
+	SetInstrumentsOptions(iopts instrument.Options) Options
+
+	RetryOptions() xretry.Options
+
+	SetRetryOptions(ropts xretry.Options) Options
+
+	WatchChanCheckInterval() time.Duration
+
+	SetWatchChanCheckInterval(t time.Duration) Options
+
+	WatchChanResetInterval() time.Duration
+
+	SetWatchChanResetInterval(t time.Duration) Options
+
+	Validate() error
+}
+
+type options struct {
+	requestTimeout         time.Duration
+	keyFn                  KeyFn
+	iopts                  instrument.Options
+	ropts                  xretry.Options
+	watchChanCheckInterval time.Duration
+	watchChanResetInterval time.Duration
+}
+
+// NewOptions creates a sane default Option
+func NewOptions() Options {
+	o := options{}
+	return o.SetRequestTimeout(defaultRequestTimeout).
+		SetInstrumentsOptions(instrument.NewOptions()).
+		SetRetryOptions(xretry.NewOptions()).
+		SetWatchChanCheckInterval(defaultWatchChanCheckInterval).
+		SetWatchChanResetInterval(defaultWatchChanResetInterval).
+		SetKeyFn(defaultKeyFn)
+}
+
+func (o options) Validate() error {
+	if o.iopts == nil {
+		return errors.New("no instrument options")
```

Comment:
yeah they are only used once here, I really don't want to assign them each with a name but only use it once unless we are really benefitting from it.